common/src/lib.rs

```diff
@@ -2,9 +2,73 @@ pub mod crypto;
 
 use crypto::*;
 use serde::{Deserialize, Serialize};
-use std::collections::BTreeSet;
+use std::collections::{BTreeMap, BTreeSet};
 use thiserror::Error;
 
+/// A set of state that is directly recorded in the header.
+///
+/// This state effects the consensus, unlike the ordinary state (which is used for data recording).
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct EssentialState {
+    /// The original validator set information before the delegation calculation.
+    ///
+    /// The order here is the priority order of the validators
+    /// which directly effects the result of `calculate_net_validator_set()`.
+    pub validator_set: Vec<(PublicKey, u64)>,
+    /// The protocol version that must be used from next block.
+    ///
+    /// It must be a valid semantic version (e.g., `0.2.3`).
+    pub version: String,
+    /// The delegation state (delegator, delegatee).
+    pub delegation: Vec<(PublicKey, PublicKey)>,
+}
+
+impl EssentialState {
+    /// Calculate the actual set of validator & voteing power for the next block.
+    ///
+    /// The order here is same as the order of leaders in each round.
+    /// returns `None` if the delegation is not valid.
+    pub fn calculate_net_validator_set(&self) -> Option<Vec<(PublicKey, u64)>> {
```

Comment:
Why do you use `Option` instead of `Result`?
I think if the delegation verification fails, the detail should be notified to the user (as an error type)

Reply:
fixed.